browse-next/src/components/DeviceSetupDefineMask.vue

```diff
@@ -0,0 +1,823 @@
+    !("x" in lastPoint)
+  ) {
+    return false;
+  }
+
+  const distance = Math.sqrt(
+    Math.pow(
+      firstPoint.x * canvas.value.width - lastPoint.x * canvas.value.width,
+      2
+    ) +
+      Math.pow(
+        firstPoint.y * canvas.value.height - lastPoint.y * canvas.value.height,
+        2
+      )
+  );
+
+  if (distance < polygonClosedTolerance.value) {
+    points.value[points.value.length - 1] = points.value[0];
+  }
+
+  clearMask(); //causes them to snap as desired, however recursion error
+  generateMask();
+  drawPolygon();
+  return distance < polygonClosedTolerance.value;
+}
+
+const toggleAreaSelect = () => {
+  updateExistingMaskRegions();
+  selectingArea.value = !selectingArea.value;
+  points.value = [];
+  clearMask();
+  generateMask();
+};
+
+const toggleMaskEnabled = () => {
+  maskEnabled.value = !maskEnabled.value;
+  if (maskEnabled.value) {
+    points.value = [];
+    clearMask();
+    generateMask();
+  } else {
+    clearMask();
+  }
+};
+
+const toggleInclusionRegion = () => {
+  inclusionRegion.value = !inclusionRegion.value;
+  generateMask();
+};
+
+function removePoint(): void {
+  if (points.value.length > 1) {
+    points.value.pop();
+    drawPolygon();
+  } else {
+    points.value = [];
+  }
+  clearMask();
+  generateMask();
+  drawPolygon();
+}
+
+function addRegionSelection(): void {
+  regionsArray.value.push({ regionData: points.value });
+  points.value = [];
+  toggleCreatingRegion();
+  const canvasElement = canvas.value;
+  const context = canvasElement.getContext("2d");
+  context.clearRect(0, 0, canvasElement.width, canvasElement.height);
+  generateMask();
+}
+
+function generateMask() {
+  polygonClosedTolerance.value = devicePixelRatio.pixelRatio.value * 10;
+  if (regionsArray.value.length === 0) {
+    return;
+  }
+  const maskCanvas = canvas.value;
+  const maskContext = maskCanvas.getContext("2d");
+  maskCanvas.width = cptvFrameWidth.value * devicePixelRatio.pixelRatio.value;
+  maskCanvas.height = cptvFrameHeight.value * devicePixelRatio.pixelRatio.value;
+
+  if (inclusionRegion.value) {
+    regionsArray.value.forEach((region) => {
+      maskContext.beginPath();
+      const regionData = region.regionData;
+      maskContext.moveTo(
+        regionData[0].x * maskCanvas.width,
+        regionData[0].y * maskCanvas.height
+      );
+      for (let i = 1; i < regionData.length; i++) {
+        maskContext.lineTo(
+          regionData[i].x * maskCanvas.width,
+          regionData[i].y * maskCanvas.height
+        );
+      }
+      maskContext.closePath();
+      maskContext.fillStyle = "rgba(0, 0, 0, 0.7)";
+      maskContext.fill("evenodd");
+    });
+  } else {
+    maskContext.beginPath();
+    maskContext.moveTo(0, 0);
+    maskContext.lineTo(maskCanvas.width, 0);
+    maskContext.lineTo(maskCanvas.width, maskCanvas.height);
+    maskContext.lineTo(0, maskCanvas.height);
+    maskContext.lineTo(0, 0);
+
+    regionsArray.value.forEach((region) => {
```

Comment:
You may want to destructure `region` when you iterate over the array: `regionsArray.value.forEach(({regionData}) => {`